email notification support

lethunder opened the pull request:
user will be able to be notify in case of a bug in the application by setting in rails_exception_handler initializer

config.email = "email1,email2"

User should then create a mailer (ErrorMailer) with the following method

  def send_error_mail_to_admin(info,email)
    @info = JSON.parse(info)
    mail(to: email, subject: 'An error occured on your application')
  end

He will also need to create views/error_mailer/send_error_mail_to_admin.html


<p>
    Hi boss,
</p>
<p>
    an error occured at <%= @info["created_at"] %>
</p>


<p>
    TARGET URL <br/>
    <%= @info["target_url"] %>
</p>


<% unless @info["referer_url"].blank? %>
  <p>
      REFERER URL <br/>
      <%= @info["referer_url"] %>
  </p>
<%end%>

<p>
    PARAMETERS <br/>
    <%= @info["params"] %>
</p>

<p>
    USER AGENT <br/>
    <%= @info["user_agent"] %>
</p>

<p>
    USER INFORMATION <br/>
    <%= @info["user_info"] %>
</p>

<p>
    CLASS NAME<br/>
    <%= @info["class_name"] %>

</p>

<% unless @info["message"].blank? %>
  <p>
      MESSAGE<br/>
      <%= @info["message"] %>
  </p>
<%end%>

<p>
    TRACE <br/>
    <%= @info["trace"] %>
</p>


<% unless @info["doc_root"].blank? %>
  <p>
      DOC ROOT <br/>
      <%= @info["doc_root"] %>
  </p>
<%end%>

Sharagoz commented:
Hi, thanks for the PR but I cant read this diff. Every line comes up as changed. Did you change the line endings from linux style to windows style? (Click the "files changed" tab and you'll see it)

I'd like the implementation changed slightly into a storage strategi that works just like activerecord/mongoid/rails_log

I havent worked on this in a while but I believe something like this would work:
`config.storage_strategies << { email:  "email1,email2" }` with a few changes to RailsExceptionHandler::Configuration.
The actual email sending would then need be moved into it's own method in RailsExceptionHandler::Configuration. This way you no longer depend on activerecord/mongoid storage being enabled to make email notifications work.

lethunder commented:
hello, thanks for your work :)

i updated the RailsExceptionHandler::Configuration class

I added :email in attr_accessor and the initialize method.

For the RailsExceptionHandler::Storage class 

i added this method
+  def self.notify(info,email)
+          ErrorMailer.send_error_mail_to_admin(info.to_json,email).deliver_later
+  end

Then i call it after error created in DB in self.active_record(info) and self.mongoid(info) with

the following lines

+	email = RailsExceptionHandler.configuration.email
+	RailsExceptionHandler::Storage.notify(info,email) unless email.blank?

Sharagoz commented:
Your implementation is 85% perfect :smiley: 
The only problem is that the current implementation means you need either activerecord storage enabled or mongoid storage enabled. The exception handler is designed in a modular way so that you can enable one or more storage strategies (there are currently 4 (activerecord, mongoid, remote_url, rails_log), email would be number 5).

If you look at how remote_url is implemented Im pretty sure email could work in exactly the same way.

Are you able to revert the changes from linux line breaks to windows  line breaks? That is something that clutters the git history and makes it harder to see when a line was last changed.

lethunder commented:
you are right! :)

i'll fix this when i'll be home. i'll be done within 3 hours
Thanks

2017-08-28 16:32 GMT+02:00 Bjørn Trondsen <notifications@github.com>:

> Your implementation is 85% perfect 😃
> The only problem is that the current implementation means you need either
> activerecord storage enabled or mongoid storage enabled. The exception
> handler is designed in a modular way so that you can enable one or more
> storage strategies (there are currently 4 (activerecord, mongoid,
> remote_url, rails_log), email would be number 5).
>
> If you look at how remote_url is implemented Im pretty sure email could
> work in exactly the same way.
>
> Are you able to revert the changes from linux line breaks to windows line
> breaks? That is something that clutters the git history and makes it harder
> to see when a line was last changed.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Sharagoz/rails_exception_handler/pull/17#issuecomment-325370365>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF7gVCvqCT3GfOq3YLs1IAG6GWEuohBUks5scs90gaJpZM4PB35S>
> .
>



-- 
Cordialement,
Didier Banlock

lethunder commented:
Will you be available in 4 hours?

2017-08-28 16:39 GMT+02:00 Didier BANLOCK <didier.banlock@gmail.com>:

> you are right! :)
>
> i'll fix this when i'll be home. i'll be done within 3 hours
> Thanks
>
> 2017-08-28 16:32 GMT+02:00 Bjørn Trondsen <notifications@github.com>:
>
>> Your implementation is 85% perfect 😃
>> The only problem is that the current implementation means you need either
>> activerecord storage enabled or mongoid storage enabled. The exception
>> handler is designed in a modular way so that you can enable one or more
>> storage strategies (there are currently 4 (activerecord, mongoid,
>> remote_url, rails_log), email would be number 5).
>>
>> If you look at how remote_url is implemented Im pretty sure email could
>> work in exactly the same way.
>>
>> Are you able to revert the changes from linux line breaks to windows line
>> breaks? That is something that clutters the git history and makes it harder
>> to see when a line was last changed.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Sharagoz/rails_exception_handler/pull/17#issuecomment-325370365>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AF7gVCvqCT3GfOq3YLs1IAG6GWEuohBUks5scs90gaJpZM4PB35S>
>> .
>>
>
>
>
> --
> Cordialement,
> Didier Banlock
>



-- 
Cordialement,
Didier Banlock

Sharagoz commented:
Thanks. Im not sure if I'll  be online in 4 hours but I check github at least once per day (mon-fri) so you wont have to wait long.
